File: C++/VendingMachine/VendingMachine/VendingMachine.cpp
#include "pch.h"
#include "VendingMachine.h"
#include "NoVendSelectionError.h"
#include <algorithm>
#include <utility>

VendingMachine::VendingMachine(Rails& rails, const Credit& credit)
	: VendingMachineRails(std::move(rails)), VendingMachineCredit(credit)
{
}

double VendingMachine::GetSelectedRailPrice()
{
	return VendingMachineRails.HasSelectedRail()
		       ? VendingMachineRails.GetSelectedRail()->GetPrice()
		       : 0;
}

void VendingMachine::SelectRail(const std::string& railCode)
{
	if (VendingMachineRails.CanSelectRail(railCode))
		VendingMachineRails.SelectRail(railCode);
}

void VendingMachine::Vend()
{
	if (!VendingMachineRails.HasSelectedRail())
		throw NoVendSelectionError();

	auto selectedRail = VendingMachineRails.GetSelectedRail();
	VendingMachineCredit.Spend(selectedRail->GetPrice());
	selectedRail->Vend();
	VendingMachineRails.ClearRailSelection();
}

bool VendingMachine::CanVend()
{
	if (!VendingMachineRails.HasSelectedRail())
		return false;
	
	const auto selectedRail = VendingMachineRails.GetSelectedRail();
	return selectedRail->CanVend()
		&& VendingMachineCredit.CanSpend(selectedRail->GetPrice());
}

void VendingMachine::AddCredit(const double amount)
{
	VendingMachineCredit.Add(amount);
}

double VendingMachine::GetCreditValue() const
{
	return VendingMachineCredit.GetValue();
}

bool VendingMachine::CanSelectRail(const std::string& railCode) const
{
	return VendingMachineRails.CanSelectRail(railCode);
}

std::map<std::string, std::string> VendingMachine::GetRailsSummary() const
{
	return VendingMachineRails.GetRailsSummary();
}

File: java/VendingMachine/src/com/ravendarque/vendingMachine/vending/VendingMachine.java
package com.ravendarque.vendingMachine.vending;

import com.ravendarque.vendingMachine.credit.Credit;
import com.ravendarque.vendingMachine.rails.Rail;
import com.ravendarque.vendingMachine.rails.Rails;

import java.util.Map;

public class VendingMachine {

    private final Rails rails;
    private final Credit credit;

    public VendingMachine(Rails rails, Credit credit) {

        this.rails = rails;
        this.credit = credit;
    }

    public void addCredit(double creditValue) {

        credit.add(creditValue);
    }

    public boolean canVend() {

        return rails.hasSelectedRail()
                && !rails.getSelectedRail()
                         .isEmpty()
                && credit.canSpend(getSelectedRailPrice());

    }

    public boolean canSelectRail(String railCode) {

        return rails.canSelectRail(railCode);
    }

    public double getCreditValue() {

        return credit.getValue();
    }

    public Map<String, String> getRailsSummary() {

        return rails.getRailsSummary();
    }

    public double getSelectedRailPrice() {

        return !rails.hasSelectedRail()
               ? 0
               : rails.getSelectedRail()
                      .getPrice();
    }

    public void selectRail(String railCode) {

        rails.selectRail(railCode);
    }

    public void vend()
            throws Exception {

        if (!rails.hasSelectedRail())
            throw new NoVendSelectionException();

        final Rail selectedRail = rails.getSelectedRail();
        credit.spend(selectedRail.getPrice());
        selectedRail.vend();
        rails.clearSelection();
    }
}
